### PR TITLE
feat: create axios instance on req.$axios for server-middleware use

### LINKED ISSUE #427: Expose axios instance for use in server middleware
I'd like to use axios instance created by this module in server middleware.

It's currently only available in webpack context and for server middleware I need to create my own which means I have to duplicate configuration and also I don't have helper methods available like `$get`.

The instance could be created directly from the module (not from the plugin as that runs later than server middlewares are parsed) and assigned to `process.axios` for example.

### PR DESCRIPTION
**Pull request by rchl**
An extra instance of axios is created and attached to "req.$axios" so
that server middlewares have access to the axios instance that has
extensions provided by this module.

The "proxyHeaders" option is ignored for that instance as it would be
both awkward to implement and potentially wrong. This is because
"proxyHeaders" is proxying requests from the request that initiated the
page request while server-middleware instance is kinda independent of
individual page requests.

Resolves #427

### CODE REVIEW

**Comment by rchl** [comment]
Also, I'm thinking maybe I should use `process.$axios` instead to make it a bit more unique.

But let me know what you think about this whole idea in general.

File: lib/module.js [hunk 1]
```diff
@@ -121,6 +122,10 @@ function axiosModule (_moduleOptions) {
   // Set _AXIOS_BASE_URL_ for dynamic SSR baseURL
   process.env._AXIOS_BASE_URL_ = options.baseURL
 
+  // Create a separate instance for server middlewares.
+  // Extended from a server-side plugin.
+  process.axios = Axios.create({ baseURL: options.baseURL })
```
Comment (pi0) [inline lib/module.js:127]: We can attach axios instance to `nuxt` (for other modules) and `req` object (for server middleware) to avoid leaking references to global

File: lib/plugin.js [hunk 2]
```diff
@@ -184,6 +185,14 @@ const setupProgress = (axios) => {
   axios.defaults.onDownloadProgress = onProgress
 }<% } %>
 
+if (process.server) {
```
Comment (pi0) [inline lib/plugin.js:188]: Why this change is necessary with this feature?

File: lib/module.js (hunk 1 above)
Comment (rchl) [reply lib/module.js:127]: To attach to `req` we would only be able to do it when a request is actually made. This here creates `process.axios` earlier on purpose so that there is no point during runtime where server middleware would see undefined `process.axios`.

File: lib/plugin.js (hunk 2 above)
Comment (rchl) [reply lib/plugin.js:188]: This is to extend the `process.axios` instance with extra methods and configuration specific to this module.

As I've hinted in the commit message, we could also extract the "extending code" into a separate file and use it both from the module and the plugin but that would require refactoring to remove the use of templates. And that means we couldn't include parts of its code conditionally, based on settings. The whole code would always be included in the bundles.

File: lib/module.js (hunk 1 above)
Comment (rchl) [reply lib/module.js:127]: Hmm, I think the current solution has a problem that in the first request to server-middleware the axios instance would not be extended yet (since the server-side plugin would run after server middleware).

So that pertains to the second issue you've made below -- we would need to probably extract the "extending" code into a separate file and call it immediately rather than relying on plugin extending it.

Comment (rchl) [reply lib/module.js:127]: And then putting in on `req` would actually be possible from the module and potentially be a better idea. It would mean that the instance of axios would be a bit harder to access (would have to be passed along various calls potentially) but workable.

File: lib/plugin.js (hunk 2 above)
Comment (pi0) [reply lib/plugin.js:188]: I think we can write node specific logic to create instance out of the plugin so we have clear control over behavior and also not depending on first SSR request to be done to make instance initialized

Comment (rchl) [reply lib/plugin.js:188]: What do you mean by creating an instance out of the plugin?

Extract the code into a separate file as I've suggested or just write extra code? We basically need this part accessible from the module also: https://github.com/nuxt-community/axios-module/blob/a3f708d4b2e2afa08dcedbfe58b7600c23635fde/lib/plugin.js#L6-L113
So are you suggesting to duplicate it or extract to a separate file or something else?

Comment (pi0) [reply lib/plugin.js:188]: We may refactor useful logic into a shared CJS file as well and use both for module and runtime to avoid duplicate code like [this](https://github.com/nuxt-community/pwa-module/blob/master/lib/meta/meta.merge.js)

Comment (rchl) [reply lib/plugin.js:188]: Yes. The only concern with that is that means the whole code will be included in the bundles rather than only code that is needed for enabled options. But I guess you are fine with that?

Comment (pi0) [reply lib/plugin.js:188]: ES6 / Module syntax is possible for nuxt modules with esm which is deprecated. We also disable esm support for typescript and test environments. When testing another project that depends on axios module, this can be problematic.

> Yes. The only concern with that is that means the whole code will be included in the bundles rather than only code that is needed for enabled options. But I guess you are fine with that?

We can mark cjs file as esm and use pure comment to allow tree-shaking per export but it shouldn't be a big deal for small shared utils.

I think ideally we should start migrating modules to typescript and proper build step to avoid such issues and also share logic in an easier way..

File: lib/module.js [hunk 3]
```diff
@@ -123,6 +130,27 @@ function axiosModule (_moduleOptions) {
...
   logger.debug(`baseURL: ${options.baseURL}`)
   logger.debug(`browserBaseURL: ${options.browserBaseURL}`)
+
+  // Create a separate axios instance on req.$axios for server middleware usage.
+  this.nuxt.hook('render:setupMiddleware', (app) => {
+    app.use((req, res, next) => {
+      const runtimeConfig = {
+        ...this.options.publicPrivateConfig,
+        ...this.options.privateRuntimeConfig
+      }
```
Comment (rchl) [inline lib/module.js:140]: Is there a better way to access runtime config from a module? Ideally with resolved values...

**Comment by rchl** [comment]
@pi0 What are your thoughts on this change? Any blocking issues? It's quite a big refactor so it will just keep getting broken by other changes if left unmerged.

**Comment by pi0** [comment]
Hi @rchl honestly more I'm thinking it might be a tricky feature to add and maintain for axios+http. Purpose of axios module  is to bind instance to renderContext not API /Config and as you mentioned most of features like proxyHeaders and progressbar integration are unused. Also usage of interceptors might be different.

Only shared parts i see are utilities and base url. Which we may use [ohmyfetch](https://github.com/nuxt-contrib/ohmyfetch) for interface similar to $get util and expose $config to either req or global.nuxt.$config

So I think would introduce global instance only with http v2 (or probably @nuxt/fetch) and this way we might have a more standardized way to make requests in config and API contexts.

**Comment by rchl** [comment]
OK, I'll close this then and wait for something better to come in the future.

For me, the main use of this would be to define interceptors for server and client in a common config. But I can deal with those in other ways.

**Comment by pi0** [comment]
@rchl Sure will keep you updated on early version.

BTW if it is about an error reporter module like sentry, would be pretty much better intercepting without directly depending on one axios/http implementation 😊

**Comment by rchl** [comment]
> BTW if it is about an error reporter module like sentry, would be pretty much better intercepting without directly depending on one axios/http implementation 😊

The main use case for me is to have better logging of failed requests in local development.
